### PR TITLE
api: Add cni chaining status in status API.

### LINKED ISSUE #16816: Make it easier to check the status of chaining mode
It would be helpful know the status of chaining mode. With this sort of visibility, it could be useful as clue to the user that something has gone wrong when they've configured Cilium with chaining. A simple field in `cilium status` for chaining status could be enough, e.g.`chaining` or `non-chaining`.

### PR DESCRIPTION
**Pull request by sayboras**
### Description

This commit is to add CNI chaining status in existing /status API in
cilium daemon. Also, simply add one new field in cilium status CLI to
reflect the same.

Fixes: #16816

Signed-off-by: Tam Mach <sayboras@yahoo.com>

### Testing

Testing was done in dev VM (cni chaining as none), and local kind cluster (cni chaining as generic-veth).

<details>
<summary>Local VM </summary>

```

vagrant@k8s1:~/go/src/github.com/cilium/cilium$ curl -s --unix-socket /var/run/cilium/cilium.sock http://127.0.0.1/v1/healthz | jq . | grep -A 2 cni-chaining
  "cni-chaining": {
    "mode": "none"
  },

vagrant@k8s1:~/go/src/github.com/cilium/cilium$ cilium status
KVStore:                Ok   Disabled
Kubernetes:             Ok   1.23 (v1.23.0) [linux/amd64]
Kubernetes APIs:        ["cilium/v2::CiliumClusterwideNetworkPolicy", "cilium/v2::CiliumEndpoint", "cilium/v2::CiliumNetworkPolicy", "cilium/v2::CiliumNode", "core/v1::Namespace", "core/v1::Node", "core/v1::Pods", "core/v1::Service", "discovery/v1::EndpointSlice", "networking.k8s.io/v1::NetworkPolicy"]
KubeProxyReplacement:   Disabled   
Host firewall:          Disabled
CNI Chaining:           none
...
```

</details>

<details>
<summary>Kind cluster with cni chaining enabled</summary>

```
$ kex cilium-v2rkz -- cilium status
Defaulted container "cilium-agent" out of: cilium-agent, mount-cgroup (init), clean-cilium-state (init)
KVStore:                Ok   Disabled
Kubernetes:             Ok   1.19 (v1.19.11) [linux/amd64]
Kubernetes APIs:        ["cilium/v2::CiliumClusterwideNetworkPolicy", "cilium/v2::CiliumEndpoint", "cilium/v2::CiliumNetworkPolicy", "cilium/v2::CiliumNode", "core/v1::Namespace", "core/v1::Node", "core/v1::Pods", "core/v1::Service", "discovery/v1beta1::EndpointSlice", "networking.k8s.io/v1::NetworkPolicy"]
KubeProxyReplacement:   Disabled   
Host firewall:          Disabled
CNI Chaining:           generic-veth
Cilium:                 Ok   1.11.90 (v1.11.90-4fd2e7cb81)
...

```

</details>

```release-note
api: Add cni chaining status in status API.
```

### CODE REVIEW

File: pkg/client/client.go [hunk 1]
```diff
@@ -310,6 +310,11 @@ func FormatStatusResponse(w io.Writer, sr *models.StatusResponse, sd StatusDetai
 		}
 		fmt.Fprintf(w, "\n")
 	}
+
+	if sr.CniChaining != nil {
```
Comment (sayboras) [inline pkg/client/client.go:314]: Feel free to let me know where cni chaining output should be in `cilium status`.

File: install/kubernetes/cilium/values.yaml [hunk 2]
```diff
@@ -315,8 +315,9 @@ cni:
 
   # -- Configure chaining on top of other CNI plugins. Possible values:
   #  - none
-  #  - generic-veth
   #  - aws-cni
+  #  - flannel
+  #  - generic-veth
```
Comment (sayboras) [inline install/kubernetes/cilium/values.yaml:320]: add flannel to allowed list.

File: install/kubernetes/cilium/templates/cilium-configmap.yaml [hunk 3]
```diff
@@ -405,6 +405,7 @@ data:
   #  - none
   #  - aws-cni
   #  - flannel
+  #  - generic-veth
```
Comment (sayboras) [inline install/kubernetes/cilium/templates/cilium-configmap.yaml:408]: add generic-veth to allowed list.

File: api/v1/openapi.yaml [hunk 4]
```diff
@@ -1778,6 +1778,9 @@ definitions:
       encryption:
         description: Status of transparent encryption
         "$ref": "#/definitions/EncryptionStatus"
+      cni-chaining:
+        description: Status of CNI chaining encryption
```
Comment (christarazi) [inline api/v1/openapi.yaml]: I assume copy-pasta :smile: 

```suggestion
        description: Status of CNI chaining
```

**Review by christarazi — APPROVED** [review]
Thanks for picking this up! LGTM with one non-blocking minor comment.

BTW, I appreciate the little snippets in your PR on how you tested the change and the output. Very helpful. :rocket:

File: pkg/client/client.go (hunk 1 above)
Comment (christarazi) [reply pkg/client/client.go:314]: Location seems fine to me. I don't think we have explicit rules on the order of things :shrug:

File: api/v1/openapi.yaml (hunk 4 above)
Comment (sayboras) [reply api/v1/openapi.yaml]: you got me :(, Ctrl C + Ctr V are my fav keys :face_exhaling:

**Review by rolinh — APPROVED** [review]

**Review by nbusseneau — APPROVED** [review]
LGTM, thanks.

**Review by qmonnet — APPROVED** [review]

**Comment by sayboras** [comment]
L4LB failure is related to outdated package in kind version.

Multicluster mesh failure seems relate to subnet quota/limit in GKE, which was happened before last month, I have reported this occurrence in testing channel in cilium slack.

https://github.com/cilium/cilium/actions/runs/1655298166

```
ERROR: (gcloud.container.clusters.create) ResponseError: code=400, message=This operation will exceed max secondary ranges per subnetwork (30) for subnet "default", consider reusing existing secondary ranges or use a different subnetwork.
```

**Review by ldelossa — APPROVED** [review]
LGTM

**Comment by nbusseneau** [comment]
> Multicluster mesh failure seems relate to subnet quota/limit in GKE, which was happened before last month, I have reported this occurrence in testing channel in cilium slack.

I have manually restarted the workflow run because that's what the Google doc recommends anyway 😩

**Comment by sayboras** [comment]
> I have manually restarted the workflow run because that's what the Google doc recommends anyway 😩

Noted and thanks, I will do the same in the future 💯

**Comment by pchaigno** [comment]
Cilium L4LB XDP test is broken and will be fixed by https://github.com/cilium/cilium/pull/18370. Other tests are passing and there shouldn't be anything L4LB-specific here. Merging.

**Pull request merged**